torch/distributed/_sharded_tensor/api.py

```diff
@@ -546,6 +553,176 @@ def sharding_spec(self) -> ShardingSpec:
+            >>> tensor = torch.arange(4, dtype=torch.int64) + 1 + 2 * rank
+            >>> tensor = torch.stack([tensor, tensor])
+            >>> tensor
+            tensor([[1, 2, 3, 4], [1, 2, 3, 4]]) # Rank 0
+            tensor([[3, 4, 5, 6], [3, 4, 5, 6]]) # Rank 1
+            tensor([[5, 6, 7, 8], [5, 6, 7, 8]]) # Rank 2
+            tensor([[7, 8, 9, 10], [7, 8, 9, 10]]) # Rank 3
+            >>> sharding_dim = 0
+            >>> spec = ChunkShardingSpec(
+                    dim=sharding_dim,
+                    placements=[
+                        "rank:0/cuda:0",
+                        "rank:1/cuda:1",
+                        "rank:2/cuda:2",
+                        "rank:3/cuda:3",
+                    ],
+                )
+            >>> st = _sharded_tensor.rand(spec, 8, 2)
+            >>> st.local_shards()[0].tensor = tensor
+            >>> sharding_dim = 1
+            >>> resharding_spec = ChunkShardingSpec(
+                    dim=sharding_dim,
+                    placements=[
+                        "rank:0/cuda:0",
+                        "rank:1/cuda:1",
+                        "rank:2/cuda:2",
+                        "rank:3/cuda:3",
+                    ],
+                )
+            >>> st.reshard(resharding_spec)
+            >>> tensor = st.local_shards()[0].tensor
+            >>> tensor
+            tensor([[1], [1], [3], [3], [5], [5], [7], [7]]) # Rank 0
+            tensor([[2], [2], [4], [4], [6], [6], [8], [8]]) # Rank 1
+            tensor([[3], [3], [5], [5], [7], [7], [9], [9]]) # Rank 2
+            tensor([[4], [4], [6], [6], [8], [8], [10], [10]]) # Rank 3
+        """
+        if (
+            not isinstance(resharding_spec, ChunkShardingSpec) or
+            not isinstance(self._sharding_spec, ChunkShardingSpec)
+        ):
+            raise NotImplementedError("Only ChunkShardingSpec supported for reshard.")
+        current_sharding_dim = int(self._sharding_spec.dim)  # type: ignore[attr-defined]
+        reshard_dim = int(resharding_spec.dim)
+        local_shard = self.local_shards()[0].tensor
+        current_rank = dist.get_rank(self._process_group)
+        world_size = dist.get_world_size(self._process_group)
+        if current_sharding_dim == reshard_dim:
+            if self._sharding_spec.placements == resharding_spec.placements:  # type: ignore[attr-defined]
+                return self
+            else:
+                input_split_sizes = [0] * world_size
+                sharded_dim_size = 0
+                split_size = get_split_size(self.size(reshard_dim), world_size)
+                for idx, placement in enumerate(self._sharding_spec.placements):  # type: ignore[attr-defined]
+                    if current_rank == placement.rank():  # type: ignore[union-attr]
+                        new_dim = resharding_spec.placements[idx].placement.rank()  # type: ignore[union-attr]
+                        input_split_sizes[new_dim] = local_shard.size(0)
```

Comment:
oh, this is kind of tricky. Because for this case, we don't any split and send the whole local shard to another rank. Like a point-to-point communication because the sharding dim is same as resharding dim. Only placements are different.